modin/backends/pandas/parsers.py

```diff
@@ -162,6 +162,46 @@ def get_dtypes(cls, dtypes_ids):
             .squeeze(axis=0)
         )
 
+    @classmethod
+    def get_dtypes_check_homogen(cls, dtypes_ids):
```

Comment:
What is the performance impact of this check?

Is it possible to enable it only when the types are not explicitly specified by the user?

Reply:
The complexity of this check linearly depends on the number of partitions/CPUs, so it shouldn't affect `read_csv` performance. Measurements results are next (no performance degradation is obtained, scalability issue for 5000x5000 dataset should be investigated separately):
```
MODIN_CPUS=8
       before           after         ratio
     [119a8be2]       [e92bfc04]
     <master>         <read_csv_warn_het_data>
          139±9ms          138±1ms     0.99  io.csv.TimeReadCsvSkiprows.time_skiprows([1000000, 10], None)
         927±20ms          908±8ms     0.98  io.csv.TimeReadCsvSkiprows.time_skiprows([5000, 5000], None)


MODIN_CPUS=112
       before           after         ratio
     [119a8be2]       [e92bfc04]
     <master>         <read_csv_warn_het_data>
          159±3ms          158±4ms     0.99  io.csv.TimeReadCsvSkiprows.time_skiprows([1000000, 10], None)
       5.66±0.01s       5.69±0.01s     1.00  io.csv.TimeReadCsvSkiprows.time_skiprows([5000, 5000], None)
```

Yes, we can enable this check only if `dtype` is not set, but even in this case partitions dtypes can be different (for example if the user defined `dtype` not for all columns with heterogeneous data), so i suggest to leave check even if `dtype` is set (since this check is cheap).

Reply:
Performance is indeed linearly dependent on the number of partitions, but the dependence of the number of partitions on the number of cores is quadratic :)

Reply:
Not sure that understood your point(
If we run something like this 
`MODIN_CPUS=8 python -c "import modin.pandas as pd; df = pd.read_csv('data_5000_5000.csv')"`
and check `partitions_dtypes` variable, it length will be equals to `MODIN_CPUS`:
```
> modin/modin/backends/pandas/parsers.py(199)get_dtypes_check_homogen()
-> for part_dtype in partitions_dtypes:
(Pdb) len(partitions_dtypes)
8
(Pdb) partitions_dtypes[0]
col0       int64
col1       int64
col2       int64
col3       int64
col4       int64
           ...
col4995    int64
col4996    int64
col4997    int64
col4998    int64
col4999    int64
Length: 5000, dtype: object
```
P.S. If quadratic appears from column partitioning, `dtypes` objects [not spitted](https://github.com/modin-project/modin/blob/master/modin/backends/pandas/parsers.py#L244-L247) even if number of columns is big enough.

Reply:
> P.S. If quadratic appears from column partitioning, dtypes objects not spitted even if number of columns is big enough.

Oh, I see.